agent/consul/state/state_store.go

```diff
@@ -152,36 +153,46 @@ type sessionCheck struct {
 }
 
 // NewStateStore creates a new in-memory state storage layer.
-func NewStateStore(gc *TombstoneGC) (*Store, error) {
+func NewStateStore(gc *TombstoneGC) *Store {
 	// Create the in-memory DB.
 	schema := stateStoreSchema()
 	db, err := memdb.NewMemDB(schema)
 	if err != nil {
-		return nil, fmt.Errorf("Failed setting up state store: %s", err)
+		// the only way for NewMemDB to error is if the schema is invalid. The
+		// scheme is static and tested to be correct, so any failure here would
+		// be a programming error, which should panic.
+		panic(fmt.Sprintf("failed to create state store: %v", err))
 	}
```

Comment:
I think this is the only plumbing change that could use some attention. I've removed the error return value from `NewStateStore` because it appears the only way for this to error is if the static schema is wrong. Since the schema is tested and static , this can only error during development, so a panic seems more appropriate and removes the need to handle errors when creating both a state.Store and an FSM.

Reply:
Yeah agreed. Even if it is somehow possible to trigger this error some other way, it's not something we test for and the net effect would be stopping the server anyway since it can't do useful work without a state store and won't retry createing one so no different to a panic really.